src/04_language.md

```diff
@@ -202,15 +202,161 @@ the process.
 
 ## Standard library traits
 
- - Drop <mark>TODO</mark>
+### `Drop` trait, the destructor
+
+Types implement the trait `std::ops::Drop` to perform some operations when the
+memory associated with a value of this type is to be reclaimed. `Drop` is the
+Rust equivalent of a destructor in C++ or a finalizer in Java.
+
+Dropping is done recursively from the outer value to the inner values.
+When a value goes out of scope (or is explicitly dropped with `std::mem::drop`),
+the value is dropped in two steps. The first step happens only if the type of
+this value implements `Drop`. It consists in calling the `drop` method on it.
+The second step consists in repeating the dropping process recursively on any
+field the value contains. Note that a `Drop` implementation is
+**only responsible for the outer value**.
+
+Another important point is that implementing `Drop` should not be systematic.
+It is only needed if the type requires some destructor logic. In fact, `Drop` is
+typically used to release external resources (network connections, files, etc.)
+or to release memory (e.g. in smart pointers such as `Box` or `Rc`).
+
+> ### Recommendation {{#check LANG-DROP | Justify `Drop` impl.}}
+>
+> In a Rust secure development, the implementation of the `std::ops::Drop` trait
+> should be justified, documented and peer-reviewed.
+
+Due to the typical use cases of `Drop`, `Drop` trait implementations are likely
```

Comment:
I don't think we can assume what a "typical use case" of `Drop` is. **The offender here is using `unsafe`**, whose usage should indeed be _justified, documented and peer-reviewed_.

I agree, however, that using `unsafe` within a `Drop` implementation must indeed be done with extreme care, since destructors are **not guaranteed to run** (_e.g._, through `mem::forget` or a reference-counted memory cycle).

If clean-up code is mandatory, lending a variable through a closure is more advisable:

```rust
pub
struct SomeRessource (
    (/* private */),
);

impl SomeRessource {
    // private
    unsafe
    fn cleanup (self: &'_ mut Self) { /* ... */ }
}
/* instead of
impl SomeRessource {
    pub
    fn new (/* args */) -> Self { /* ... */ }
}
impl Drop for SomeRessource {
    fn drop (self: &'_ mut Self)
    {
        unsafe { self.cleanup() }
    }
}
do */
impl SomeRessource {
    // PRIVATE !!
    fn new (/* args */) -> Self { /* ... */ }

    pub
    fn with_new<R, F> (/* args ,*/ f: F) -> R
    where
        F : FnOnce(&mut Self) -> R,
    {
        // scopeguard for exception safety
        let mut guard = ::scopeguard::guard(
            Self::new(/* args */),
            // code called when the guard is dropped
            |mut this| unsafe {
                this.cleanup();
            },
        );
        let this: &mut Self = &mut *guard;
        f(this)
    }
}
```

  - (although it does not protect against `f` never returning (infinite `loop` or `abort`)

Reply:
As for assuming a typical use case for `Drop` , I'm pretty sure it is reasonable (I'm not the only one, e.g. the [book](https://doc.rust-lang.org/book/ch15-03-drop.html)). But it might seem well... cavalier. So I will try to rewrite that.

As for the offender being unsafe, it is not always the case. `Drop` attracts bugs, even pure logic ones that require no memory safety violation at all. 

Destructors not running is indeed the main problem that **should be addressed further.** You make a very good point with your example. It might be one of the best approach for std-based programs (`scopeguard` relies on `Drop` and more problematic panic unwinding).